File: server/app/pulse.py
"""Remote photoplethysmography (rPPG) liveness — the 3-D-mask counter-measure.

Every layer before this one is defeated by a good silicone mask: it turns,
blinks through eye holes, is three-dimensional, and reflects the screen flash
like skin. What a mask does **not** have is blood. Skin colour pulses with each
heartbeat — a fraction of a percent, invisible to the eye, but measurable as
the mean colour of a skin patch over a few seconds of video. This is the same
signal a pulse-oximeter reads, recovered from a camera, and the technique the
3-D-mask PAD literature converged on (Liu, Yuen et al. 2016; Nowara 2017;
Heusch & Marcel 2019).

Pure and numeric: ``pulse_liveness_score`` takes ``(t_ms, [patch RGB…])`` per
frame — one mean colour per skin patch (forehead, both cheeks) — and returns a
score, a beats-per-minute estimate and the spectral prominence. Measuring the
patch colours from pixels is the caller's job (`services.verification`, via
`geometry.skin_patch_colors` on MediaPipe's landmarks).

Method:

1. resample the irregular phone timestamps onto a uniform grid;
2. per patch, POS (Wang, den Brinker, Stuijk & de Haan, IEEE TBME 2017):
   in a sliding window, normalise each channel by its mean, project onto the
   plane orthogonal to skin tone — ``S1 = G − B``, ``S2 = −2R + G + B`` — and
   combine ``h = S1 + (σ1/σ2)·S2``; overlap-add the windows;
3. Hann-windowed power spectrum in the physiological band 0.7–3 Hz
   (42–180 bpm), normalised per patch and **averaged across patches** — a real
   pulse peaks at the same frequency in every patch, noise does not;
4. prominence = mean power within ±0.15 Hz of the peak over the median power
   elsewhere in the band, in dB. A pulse concentrates power at one frequency; a
   mask, a photo or noise spreads it flat.

Calibrated on synthetic traces only (see test_pulse.py): with a pulse
amplitude at least twice the per-frame colour noise, real p10 ≈ 9 dB against
spoof max ≈ 7 dB; at amplitude ≈ noise the two overlap. That is why the gate
ships **advisory** (`Thresholds.pulse_rule`) until it has been measured on
phones, and why the burst is 7–8 s at the highest snapshot rate the device
gives. This layer raises the cost of a mask attack; it is not a certificate.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

#: Physiological band: 42–180 bpm.
BAND_HZ = (0.7, 3.0)
#: Half-width of the window counted as "the pulse" around the spectral peak.
PEAK_HALF_WIDTH_HZ = 0.15
#: POS window length in seconds (the paper uses 1.6 s at 20 fps).
POS_WINDOW_S = 1.6
#: A burst shorter than this cannot resolve the band; the score is 0.
MIN_FRAMES = 24
MIN_SPAN_MS = 3_000
#: Uniform resampling rate bounds. Below ~5 Hz the band aliases; above 60 is waste.
RESAMPLE_HZ = (5.0, 60.0)
#: Score = logistic(prominence_dB − centre)/scale. 7 dB → 0.5.
PROMINENCE_CENTRE_DB = 7.0
PROMINENCE_SCALE_DB = 1.0


@dataclass(slots=True)
class PulseResult:
    #: 0..1 — confidence that a periodic pulse is present.
    score: float
    #: Beats per minute at the spectral peak (0 when nothing was measured).
    bpm: float
    #: Peak prominence in dB (peak window mean vs median of the rest of the band).
    prominence_db: float
    frames: int
    span_ms: int
    sampling_hz: float
    patches: int = 0
    #: Why the score is 0 without measurement, or "" when it was measured.
    note: str = ""


def pulse_liveness_score(
    times_ms: list[int] | np.ndarray,
    colors: list | np.ndarray,
) -> PulseResult:
    """``colors`` is (n, 3) for one patch or (n, k, 3) for k patches per frame."""
    t = np.asarray(times_ms, dtype=np.float64).reshape(-1)
    c = np.asarray(colors, dtype=np.float64)
    n = int(t.shape[0])
    if c.ndim == 2:
        c = c[:, None, :]
    if n == 0 or c.ndim != 3 or c.shape[0] != n or c.shape[2] != 3:
        return PulseResult(0.0, 0.0, 0.0, n, 0, 0.0, 0, "shape")
    k = int(c.shape[1])

    order = np.argsort(t, kind="stable")
    t, c = t[order], c[order]
    keep = np.concatenate([[True], np.diff(t) > 0])  # drop duplicate timestamps
    t, c = t[keep], c[keep]
    n = int(t.shape[0])
    span = int(round(t[-1] - t[0])) if n > 1 else 0
    if n < MIN_FRAMES or span < MIN_SPAN_MS:
        return PulseResult(0.0, 0.0, 0.0, n, span, 0.0, k, "too_short")
    if not np.all(np.isfinite(c)):
        return PulseResult(0.0, 0.0, 0.0, n, span, 0.0, k, "nan")

    # 1. uniform resampling at the burst's own median rate
    median_dt = float(np.median(np.diff(t))) / 1000.0
    fs = float(np.clip(1.0 / max(median_dt, 1e-3), *RESAMPLE_HZ))
    grid = np.arange(t[0], t[-1], 1000.0 / fs)
    if grid.shape[0] < MIN_FRAMES:
        return PulseResult(0.0, 0.0, 0.0, n, span, fs, k, "too_short")

    # 2./3. per-patch POS → normalised band spectrum, averaged
    accumulated: np.ndarray | None = None
    freqs: np.ndarray | None = None
    used = 0
    for patch in range(k):
        trace = c[:, patch, :]
        if np.any(trace.std(axis=0) < 1e-9):
            continue  # a constant patch (photo, saturated) carries nothing
        rgb = np.stack([np.interp(grid, t, trace[:, ch]) for ch in range(3)], axis=1)
        signal = pos_signal(rgb, fs)
        if signal is None or signal.std() < 1e-12:
            continue
        f, power = band_spectrum(signal, fs)
        total = float(power.sum())
        if total <= 0.0:
            continue
        power = power / total
        accumulated = power if accumulated is None else accumulated + power
        freqs = f
        used += 1
    if accumulated is None or freqs is None:
        return PulseResult(0.0, 0.0, 0.0, n, span, fs, k, "flat")
    accumulated /= used

    # 4. prominence of the strongest peak
    prominence_db, peak_hz = peak_prominence(freqs, accumulated)
    score = float(1.0 / (1.0 + np.exp(-(prominence_db - PROMINENCE_CENTRE_DB) / PROMINENCE_SCALE_DB)))
    return PulseResult(score, float(peak_hz * 60.0), float(prominence_db), n, span, fs, used)


def pos_signal(rgb: np.ndarray, fs: float) -> np.ndarray | None:
    """Plane-Orthogonal-to-Skin pulse signal from a uniformly sampled RGB trace."""
    n = rgb.shape[0]
    win = max(int(round(POS_WINDOW_S * fs)), 8)
    if n < win:
        win = n
    h = np.zeros(n, dtype=np.float64)
    projection = np.array([[0.0, 1.0, -1.0], [-2.0, 1.0, 1.0]])
    for start in range(0, n - win + 1):
        block = rgb[start : start + win]
        mean = block.mean(axis=0)
        if np.any(mean <= 1e-9):
            continue
        normalised = block / mean - 1.0
        s = normalised @ projection.T  # (win, 2)
        s1, s2 = s[:, 0], s[:, 1]
        sigma2 = s2.std()
        alpha = s1.std() / sigma2 if sigma2 > 1e-12 else 0.0
        p = s1 + alpha * s2
        h[start : start + win] += p - p.mean()
    return h if np.any(h != 0.0) else None


def band_spectrum(signal: np.ndarray, fs: float) -> tuple[np.ndarray, np.ndarray]:
    """(frequencies, power) restricted to the physiological band."""
    x = signal - signal.mean()
    n = x.shape[0]
    window = np.hanning(n)
    nfft = int(2 ** np.ceil(np.log2(max(n, 8) * 8)))  # zero-pad for a fine grid
    spectrum = np.abs(np.fft.rfft(x * window, nfft)) ** 2
    freqs = np.fft.rfftfreq(nfft, d=1.0 / fs)
    band = (freqs >= BAND_HZ[0]) & (freqs <= BAND_HZ[1])
    return freqs[band], spectrum[band]


def peak_prominence(freqs: np.ndarray, power: np.ndarray) -> tuple[float, float]:
    """(prominence in dB, peak Hz): mean power around the peak over the median
    power of the rest of the band (first harmonic excluded from "rest")."""
    if freqs.size == 0:
        return -30.0, 0.0
    peak_index = int(np.argmax(power))
    peak_hz = float(freqs[peak_index])
    in_peak = np.abs(freqs - peak_hz) <= PEAK_HALF_WIDTH_HZ
    in_harmonic = np.abs(freqs - 2.0 * peak_hz) <= PEAK_HALF_WIDTH_HZ
    rest = power[~(in_peak | in_harmonic)]
    if rest.size == 0:
        return 30.0, peak_hz
    floor = float(np.median(rest))
    if floor <= 1e-18:
        return 30.0, peak_hz
    prominence = 10.0 * np.log10(float(power[in_peak].mean()) / floor)
    return float(np.clip(prominence, -30.0, 30.0)), peak_hz

File: packages/react-native-ekyc-local/src/pulse.ts
/**
 * rPPG pulse liveness on the phone — the silicone-mask counter-measure, in
 * pure TypeScript. A line-for-line port of the server's `app/pulse.py`
 * (POS projection per skin patch, Hann-windowed spectrum in 0.7–3 Hz,
 * spectra averaged across patches, peak prominence in dB → logistic score),
 * so the two agree on the same input.
 *
 * Skin has blood; masks, prints and screens showing a still image do not.
 * The face's mean colour over a few seconds carries the heartbeat as a tiny
 * periodic change. This module only does the maths; sampling the patch colours
 * from frames is `PulseSampler`'s job in `LocalLivenessCamera`.
 *
 * Calibrated on synthetic traces only (see tests): clean separation when the
 * pulse amplitude is ≥ ~2× the per-frame colour noise, overlap at ≈1×. Hence
 * `pulseRule` defaults to `advisory` — the score is reported, not enforced —
 * until real phones have been measured.
 */

export const BAND_HZ: [number, number] = [0.7, 3.0]
export const PEAK_HALF_WIDTH_HZ = 0.15
export const POS_WINDOW_S = 1.6
export const MIN_FRAMES = 24
export const MIN_SPAN_MS = 3000
export const RESAMPLE_HZ: [number, number] = [5, 60]
export const PROMINENCE_CENTRE_DB = 7.0
export const PROMINENCE_SCALE_DB = 1.0
/** Default gate on `score` (logistic of prominence, 0.5 at 7 dB). */
export const DEFAULT_PULSE_MIN = 0.5

export type Rgb = [number, number, number]

export type PulseResult = {
  /** 0..1 — confidence that a periodic pulse is present. */
  score: number
  bpm: number
  prominenceDb: number
  frames: number
  spanMs: number
  samplingHz: number
  patches: number
  /** Why the score is 0 without measurement, or '' when measured. */
  note: '' | 'shape' | 'too_short' | 'nan' | 'flat'
}

/** `colors[i]` = one RGB (0..1) per patch for frame i. */
export function pulseLivenessScore(timesMs: ArrayLike<number>, colors: Rgb[][]): PulseResult {
  const n0 = timesMs.length
  const k = colors[0]?.length ?? 0
  if (n0 === 0 || colors.length !== n0 || k === 0) return zero(n0, 0, 0, k, 'shape')

  // sort by time, drop duplicate timestamps
  const order = Array.from({ length: n0 }, (_, i) => i).sort((a, b) => timesMs[a]! - timesMs[b]!)
  const t: number[] = []
  const c: Rgb[][] = []
  for (const i of order) {
    const ti = timesMs[i]!
    if (t.length > 0 && ti <= t[t.length - 1]!) continue
    const row = colors[i]!
    if (row.length !== k) return zero(n0, 0, 0, k, 'shape')
    t.push(ti)
    c.push(row)
  }
  const n = t.length
  const span = n > 1 ? Math.round(t[n - 1]! - t[0]!) : 0
  if (n < MIN_FRAMES || span < MIN_SPAN_MS) return zero(n, span, 0, k, 'too_short')
  for (const row of c) for (const p of row) for (const v of p) if (!Number.isFinite(v)) return zero(n, span, 0, k, 'nan')

  // 1. uniform resampling at the burst's own median rate
  const dts: number[] = []
  for (let i = 1; i < n; i++) dts.push(t[i]! - t[i - 1]!)
  const medianDt = median(dts) / 1000
  const fs = clamp(1 / Math.max(medianDt, 1e-3), RESAMPLE_HZ[0], RESAMPLE_HZ[1])
  const grid: number[] = []
  for (let g = t[0]!; g < t[n - 1]!; g += 1000 / fs) grid.push(g)
  if (grid.length < MIN_FRAMES) return zero(n, span, fs, k, 'too_short')

  // 2./3. per-patch POS → normalised band spectrum, averaged
  let accumulated: Float64Array | null = null
  let freqs: Float64Array | null = null
  let used = 0
  for (let patch = 0; patch < k; patch++) {
    const channels: number[][] = [0, 1, 2].map((ch) => c.map((row) => row[patch]![ch]!))
    if (channels.some((ch) => std(ch) < 1e-9)) continue
    const rgb = channels.map((ch) => interp(grid, t, ch))
    const signal = posSignal(rgb, fs)
    if (!signal || std(Array.from(signal)) < 1e-12) continue
    const { freqs: f, power } = bandSpectrum(signal, fs)
    let total = 0
    for (const v of power) total += v
    if (total <= 0) continue
    for (let i = 0; i < power.length; i++) power[i] = power[i]! / total
    if (!accumulated) {
      accumulated = new Float64Array(power)
      freqs = f
    } else {
      for (let i = 0; i < power.length; i++) accumulated[i] = accumulated[i]! + power[i]!
    }
    used++
  }
  if (!accumulated || !freqs) return zero(n, span, fs, k, 'flat')
  for (let i = 0; i < accumulated.length; i++) accumulated[i] = accumulated[i]! / used

  // 4. prominence of the strongest peak
  const { prominenceDb, peakHz } = peakProminence(freqs, accumulated)
  const score = 1 / (1 + Math.exp(-(prominenceDb - PROMINENCE_CENTRE_DB) / PROMINENCE_SCALE_DB))
  return { score, bpm: peakHz * 60, prominenceDb, frames: n, spanMs: span, samplingHz: fs, patches: used, note: '' }
}

function zero(frames: number, spanMs: number, fs: number, patches: number, note: PulseResult['note']): PulseResult {
  return { score: 0, bpm: 0, prominenceDb: 0, frames, spanMs, samplingHz: fs, patches, note }
}

/** Plane-Orthogonal-to-Skin pulse signal from uniformly sampled RGB channels. */
export function posSignal(rgb: number[][], fs: number): Float64Array | null {
  const n = rgb[0]!.length
  let win = Math.max(Math.round(POS_WINDOW_S * fs), 8)
  if (n < win) win = n
  const h = new Float64Array(n)
  let any = false
  for (let start = 0; start + win <= n; start++) {
    const mean = [0, 1, 2].map((ch) => {
      let s = 0
      for (let i = start; i < start + win; i++) s += rgb[ch]![i]!
      return s / win
    })
    if (mean.some((m) => m <= 1e-9)) continue
    const s1 = new Float64Array(win)
    const s2 = new Float64Array(win)
    for (let i = 0; i < win; i++) {
      const r = rgb[0]![start + i]! / mean[0]! - 1
      const g = rgb[1]![start + i]! / mean[1]! - 1
      const b = rgb[2]![start + i]! / mean[2]! - 1
      s1[i] = g - b
      s2[i] = -2 * r + g + b
    }
    const sigma2 = std(Array.from(s2))
    const alpha = sigma2 > 1e-12 ? std(Array.from(s1)) / sigma2 : 0
    const p = new Float64Array(win)
    let pm = 0
    for (let i = 0; i < win; i++) {
      p[i] = s1[i]! + alpha * s2[i]!
      pm += p[i]!
    }
    pm /= win
    for (let i = 0; i < win; i++) {
      h[start + i] = h[start + i]! + (p[i]! - pm)
      any = true
    }
  }
  return any ? h : null
}

/** (freqs, power) restricted to the physiological band — Hann window, zero-padded DFT. */
export function bandSpectrum(signal: Float64Array, fs: number): { freqs: Float64Array; power: Float64Array } {
  const n = signal.length
  let mean = 0
  for (const v of signal) mean += v
  mean /= n
  const nfft = 1 << Math.ceil(Math.log2(Math.max(n, 8) * 8))
  const x = new Float64Array(nfft)
  for (let i = 0; i < n; i++) x[i] = (signal[i]! - mean) * (0.5 - 0.5 * Math.cos((2 * Math.PI * i) / (n - 1)))
  const df = fs / nfft
  const kLo = Math.ceil(BAND_HZ[0] / df)
  const kHi = Math.floor(BAND_HZ[1] / df)
  const count = Math.max(0, kHi - kLo + 1)
  const freqs = new Float64Array(count)
  const power = new Float64Array(count)
  // Direct DFT over the band bins only — the band is narrow (≈2.3 Hz of a
  // ≤60 Hz spectrum) so this is far cheaper than a full FFT and needs no library.
  for (let j = 0; j < count; j++) {
    const kk = kLo + j
    let re = 0
    let im = 0
    const w = (-2 * Math.PI * kk) / nfft
    for (let i = 0; i < n; i++) {
      const a = w * i
      re += x[i]! * Math.cos(a)
      im += x[i]! * Math.sin(a)
    }
    freqs[j] = kk * df
    power[j] = re * re + im * im
  }
  return { freqs, power }
}

export function peakProminence(freqs: Float64Array, power: Float64Array): { prominenceDb: number; peakHz: number } {
  if (freqs.length === 0) return { prominenceDb: -30, peakHz: 0 }
  let peak = 0
  for (let i = 1; i < power.length; i++) if (power[i]! > power[peak]!) peak = i
  const peakHz = freqs[peak]!
  const rest: number[] = []
  let inPeakSum = 0
  let inPeakN = 0
  for (let i = 0; i < freqs.length; i++) {
    const f = freqs[i]!
    const inPeak = Math.abs(f - peakHz) <= PEAK_HALF_WIDTH_HZ
    const inHarm = Math.abs(f - 2 * peakHz) <= PEAK_HALF_WIDTH_HZ
    if (inPeak) {
      inPeakSum += power[i]!
      inPeakN++
    }
    if (!inPeak && !inHarm) rest.push(power[i]!)
  }
  if (rest.length === 0) return { prominenceDb: 30, peakHz }
  const floor = median(rest)
  if (floor <= 1e-18) return { prominenceDb: 30, peakHz }
  const prom = 10 * Math.log10(inPeakSum / inPeakN / floor)
  return { prominenceDb: clamp(prom, -30, 30), peakHz }
}

// ---- small numeric helpers ---------------------------------------------------

function median(values: number[]): number {
  if (values.length === 0) return 0
  const s = [...values].sort((a, b) => a - b)
  const m = s.length >> 1
  return s.length % 2 ? s[m]! : (s[m - 1]! + s[m]!) / 2
}
function std(values: ArrayLike<number>): number {
  const n = values.length
  if (n === 0) return 0
  let m = 0
  for (let i = 0; i < n; i++) m += values[i]!
  m /= n
  let v = 0
  for (let i = 0; i < n; i++) v += (values[i]! - m) ** 2
  return Math.sqrt(v / n)
}
function clamp(v: number, lo: number, hi: number): number {
  return Math.min(hi, Math.max(lo, v))
}
/** Linear interpolation of (xs, ys) at each grid point; xs strictly increasing. */
function interp(grid: number[], xs: number[], ys: number[]): number[] {
  const out: number[] = new Array(grid.length)
  let j = 0
  for (let i = 0; i < grid.length; i++) {
    const g = grid[i]!
    while (j < xs.length - 2 && xs[j + 1]! < g) j++
    const x0 = xs[j]!
    const x1 = xs[j + 1]!
    const y0 = ys[j]!
    const y1 = ys[j + 1]!
    out[i] = x1 === x0 ? y0 : y0 + ((y1 - y0) * (g - x0)) / (x1 - x0)
  }
  return out
}
